presto-main/src/main/java/io/prestosql/operator/scalar/RowToJsonCast.java

```diff
@@ -75,18 +82,43 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         checkCondition(canCastToJson(type), INVALID_CAST_ARGUMENT, "Cannot cast %s to JSON", type);
 
         List<Type> fieldTypes = type.getTypeParameters();
+        List<TypeSignatureParameter> typeSignatureParameters = type.getTypeSignature().getParameters();
+
+        List<String> fieldNames = new ArrayList<>(fieldTypes.size());
         List<JsonGeneratorWriter> fieldWriters = new ArrayList<>(fieldTypes.size());
         for (int i = 0; i < fieldTypes.size(); i++) {
-            fieldWriters.add(createJsonGeneratorWriter(fieldTypes.get(i)));
+            fieldNames.add(typeSignatureParameters.get(i).getNamedTypeSignature().getName().orElse("field" + i));
+            fieldWriters.add(createJsonGeneratorWriter(fieldTypes.get(i), legacyRowToJson));
         }
-        MethodHandle methodHandle = METHOD_HANDLE.bindTo(fieldWriters);
+        MethodHandle methodHandle = legacyRowToJson ? LEGACY_METHOD_HANDLE.bindTo(fieldWriters) : METHOD_HANDLE.bindTo(fieldNames).bindTo(fieldWriters);
 
         return new ScalarFunctionImplementation(
                 false,
                 ImmutableList.of(valueTypeArgumentProperty(RETURN_NULL_ON_NULL)),
                 methodHandle);
     }
 
+    @UsedByGeneratedCode
+    public static Slice toJson(List<String> fieldNames, List<JsonGeneratorWriter> fieldWriters, ConnectorSession session, Block block)
```

Comment:
rename to `toJsonObject`. I did not spot the difference between first and second method at first read.